### DISCUSSION TITLE
`free` instruction

### DISCUSSION THREAD

**Discussion by MinekPo1** [Its-Jakey/SCPP#3] [Ideas]
Please. Since there is a limited amount of ram on the VM memory management has no value if you cannot free memory.

Take this program for example:

```SCPP

namespace Program{
    func A(){
        var arr[400000];
    }

    func B(){
        var arr[400000];
    }

    func main(){
        A();
        B();
    }
}
```

The total amount of ram used at any moment is never to large for the VM. However since `A`'s `arr` cannot be de-allocated the VM crashes since it is unable to allocate a total of 800000 cells of memory.

Of course, a smart compiler could manually reuse space, but that's difficult and why.

**Comment by Its-Jakey** [comment]
> Please. Since there is a limited amount of ram on the VM memory management has no value if you cannot free memory.
> 
> 
> 
> Take this program for example:
> 
> 
> 
> ```SCPP
> 
> 
> 
> namespace Program{
> 
>     func A(){
> 
>         var arr[400000];
> 
>     }
> 
> 
> 
>     func B(){
> 
>         var arr[400000];
> 
>     }
> 
> 
> 
>     func main(){
> 
>         A();
> 
>         B();
> 
>     }
> 
> }
> 
> ```
> 
> 
> 
> The total amount of ram used at any moment is never to large for the VM. However since `A`'s `arr` cannot be de-allocated the VM crashes since it is unable to allocate a total of 800000 cells of memory.
> 
> 
> 
> Of course, a smart compiler could manually reuse space, but that's difficult and why.

I am currently working on a large commit that is almost done, I will work on adding 'free' also. Since the VM uses variables instead of addresses it's hard to implement because the VM only increments the memory pointer for each variable, and does not scan for an empty memory address.

**Reply by MinekPo1** [reply]
You can see how I implemented allocation in my implementation of the VM.

**Reply by Its-Jakey** [reply]
> You can see how I implemented allocation in my implementation of the VM.

Thanks, I'll check yours out

**Reply by Its-Jakey** [reply]
I looked through your implementation's memory management and I have no idea how I would do that on Scratch. I think I'll just write it in java (the language I know best) so I can better visualize it, then port it to scratch.
